File: algorithms/cpp/containVirus/containVirus.cpp
// Source: https://leetcode.com/problems/contain-virus/
// Author: Miao Zhang
// Date:   2021-03-06

class Solution {
public:
    int containVirus(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int res = 0;
        while (true) {
            vector<int> visited(m * n, 0);
            vector<unordered_set<int>> regions;
            vector<unordered_set<int>> frontiers;
            vector<int> walls;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    int key = i * n + j;
                    if (grid[i][j] == 1 && visited[key] == 0) {
                        regions.push_back(unordered_set<int>());
                        frontiers.push_back(unordered_set<int>());
                        walls.push_back(0);
                        dfs(i, j, visited, regions, frontiers, walls, grid);
                    }
                }
            }
            if (regions.empty()) break;

            int maxidx = 0;
            int maxlen = 0;
            for (int i = 0; i < frontiers.size(); i++) {
                if (frontiers[i].size() > maxlen) {
                    maxlen = frontiers[i].size();
                    maxidx = i;
                }
            }
            res += walls[maxidx];
            for (int i = 0; i < regions.size(); i++) {
                if (i == maxidx) {
                    for (auto r: regions[i]) {
                        grid[r / n][r % n] = 2;
                    }
                } else {
                    for (auto r: regions[i]) {
                        int x = r / n;
                        int y = r % n;
                        int dirs[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
                        for (auto d: dirs) {
                            int nx = x + d[0];
                            int ny = y + d[1];
                            if (nx >= 0 && nx < grid.size() && ny >= 0 && ny < grid[0].size() && grid[nx][ny] == 0) {
                                grid[nx][ny] = 1;
                            }
                        }
                    }
                }
            }
        }
        return res;
    }

private:
    void dfs(int i, int j, vector<int>& visited,
             vector<unordered_set<int>>& regions,
             vector<unordered_set<int>>& frontiers, vector<int>& walls,
             vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int key = i * n + j;
        if (visited[key] == 1) return;
        visited[key] = 1;
        regions.back().insert(key);
        vector<vector<int>> dirs{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (auto d: dirs) {
            int x = i + d[0];
            int y = j + d[1];
            if (x >= 0 && x < m && y >= 0 && y < n) {
                if (grid[x][y] == 1) {
                    dfs(x, y, visited, regions, frontiers, walls, grid);
                } else if(grid[x][y] == 0) {
                    frontiers.back().insert(x * n + y);
                    walls.back()++;
                }
            }
        }
    }
};

File: algorithms/python/containVirus/containVirus.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/contain-virus/
# Author: Miao Zhang
# Date:   2021-03-06

class Solution:
    def containVirus(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        
        def neighbors(i, j):
            dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            for d in dirs:
                x = i + d[0]
                y = j + d[1]
                if 0 <= x < m and 0 <= y < n:
                    yield x, y
        
        def dfs(i, j):
            if (i, j) not in visited:
                visited.add((i, j))
                regions[-1].add((i, j))
                for x, y in neighbors(i, j):
                    if grid[x][y] == 1:
                        dfs(x, y)
                    elif grid[x][y] == 0:
                        frontiers[-1].add((x, y))
                        walls[-1] += 1
        
        res = 0
        while True:
            visited = set()
            regions = [] #  Virus infected area
            frontiers = [] # Virus free zone
            walls = []
            for i in range(m):
                for j in range(n):
                    if grid[i][j] == 1 and (i, j) not in visited:
                        regions.append(set())
                        frontiers.append(set())
                        walls.append(0)
                        dfs(i, j)
            if not regions: break
            
            target_idx = frontiers.index(max(frontiers, key=len))
            res += walls[target_idx]
            for i, region in enumerate(regions):
                if i == target_idx:
                    for r, c in region:
                        grid[r][c] = 2 # protected by wall
                else:
                    for r, c in region:
                        for nx, ny in neighbors(r, c):
                            if grid[nx][ny] == 0:
                                grid[nx][ny] = 1
        
        return res
